### ISSUE TITLE
Multi-GPU simulation shows slow-down than 1-GPU simulation

### ISSUE THREAD

**Issue by sha256feng** [openmm/openmm#3548]
Hi,

I am testing OpenMM multi-GPU performance. The result shows that simulation speed using 2 GPUs (5.80 ns/day) is slower than 1 GPU (6.75 ns/day). We use the NVIDIA RTX A40 GPUs and the jobs were submitted to HPC with the Slurm job scheduler. I've checked the previous threads #2971 and #2533. Looks like, in #2971, the multi-GPU benchmarking was done on AWS, different from our case with Slurm. 

Summary of information:
- GPU: RTX A40
- Platform: CentOS 8, Slurm
- OpenMM version: 7.7.0 with cuda 11.6
- System size: 570K atoms

The OpenMM scripts and files were downloaded from CHARMM-GUI Membrane Builder, except the line `prop = dict()` was modified to add `CudaDeviceIndex='0,1'` to the dictionary definition. In Slurm, no matter which GPUs we are allocated (eg, 3,5,6), the GPU indices visible inside that Slurm environment always start from 0, according to our HPC manager. 

Here is the device setting in the OpenMM script. The whole OpenMM configuration file can be assessed [here](https://gist.github.com/sha256feng/ce82e8228ed1aa3e1247aa44185696bf)  

```
# Set platform
DEFAULT_PLATFORMS = 'CUDA', 'OpenCL', 'CPU'
enabled_platforms = [Platform.getPlatform(i).getName() for i in range(Platform.getNumPlatforms())]
if args.platform:
    if not args.platform[0] in enabled_platforms:
        print("Unable to find OpenMM platform '{}'; exiting".format(args.platform[0]), file=sys.stderr)
        sys.exit(1)

    platform = Platform.getPlatformByName(args.platform[0])
else:
    for platform in DEFAULT_PLATFORMS:
        if platform in enabled_platforms:
            platform = Platform.getPlatformByName(platform)
            break
    if isinstance(platform, str):
        print("Unable to find any OpenMM platform; exiting".format(args.platform[0]), file=sys.stderr)
        sys.exit(1)

print("Using platform:", platform.getName())
prop = dict(CudaPrecision='single', CudaDeviceIndex='0,1') if platform.getName() == 'CUDA' else dict()

# Build simulation context
simulation = Simulation(top.topology, system, integrator, platform, prop)
```

The Slurm submission script (I omitted details about cuda, intel package loading, and Python package address linking, but they are  correct for OpenMM7.7.0 with CUDA version 11.6)
```
#!/bin/csh
#SBATCH --partition=ima40-gpu
#SBATCH --gres=gpu:2
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=4
#SBATCH --export=ALL
#SBATCH -t 3:00:00

python -u openmm_run_2gpu.py -i step5_input.inp -t toppar.str -p step5_input.psf -c step5_input.crd -irst step7_1.rst -orst step7_2.rst -odcd step7_2.dcd | tee step7_2.out > /dev/null
```

Any suggestions on this? 

Thanks a lot,
Shasha

**Comment by dmclark17** [comment]
Do you know how the GPUs are connected? Running `nvidia-smi topo -m` should print the topology. 

I ran the amber20-cellulose (408,609 atoms) benchmark via the benchmark.py script in the examples directory on an A40 node and saw the following performance:

| Number of GPUs  | ns/day |
| ------------- | ------------- |
| 1 | 57.1928 |
| 2 | 91.3952 |
| 4 | 103.521 |

Is the single GPU performance you're observing in line with your expectation for the system?

**Comment by sha256feng** [comment]
Looks that we don't have NVLinks between GPUs:
```
nvidia-smi topo -m
	GPU0	GPU1	GPU2	GPU3	GPU4	GPU5	GPU6	GPU7	CPU Affinity	NUMA Affinity
GPU0	 X 	NODE	NODE	NODE	SYS	SYS	SYS	SYS	0-15	0
GPU1	NODE	 X 	NODE	NODE	SYS	SYS	SYS	SYS	0-15	0
GPU2	NODE	NODE	 X 	NODE	SYS	SYS	SYS	SYS	0-15	0
GPU3	NODE	NODE	NODE	 X 	SYS	SYS	SYS	SYS	0-15	0
GPU4	SYS	SYS	SYS	SYS	 X 	NODE	NODE	NODE	16-31	1
GPU5	SYS	SYS	SYS	SYS	NODE	 X 	NODE	NODE	16-31	1
GPU6	SYS	SYS	SYS	SYS	NODE	NODE	 X 	NODE	16-31	1
GPU7	SYS	SYS	SYS	SYS	NODE	NODE	NODE	 X 	16-31	1

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks
```

The speed on a single GPU is comparable to yours, 57.17 ns/day :)
```
python share/openmm/examples/benchmark.py --platform CUDA --test amber20-cellulose --device=0
Platform: CUDA
Precision: single
Device: 0

Test: amber20-cellulose
Ensemble: NVT
Downloading https://ambermd.org/Amber20_Benchmark_Suite.tar.gz
Extracting Amber20_Benchmark_Suite.tar.gz
Step Size: 4 fs
Integrated 9831 steps in 59.426 seconds
57.1735 ns/day
```

**Comment by sha256feng** [comment]
Yes, I am able to do that. The performance on two GPUs is again similar to yours.  
```
python share/openmm/examples/benchmark.py --platform CUDA --test amber20-cellulose --device=0,1
Platform: CUDA
Precision: single
Device: 0,1

Test: amber20-cellulose
Ensemble: NVT
Step Size: 4 fs
Integrated 15941 steps in 59.9895 seconds
91.8362 ns/day
```
So the problem probably has something to do with our system or the CHARMM-GUI-generated settings?

**Comment by dmclark17** [comment]
That would be my guess—this is beyond my area of expertise. @peastman do you have any ideas?

**Comment by peastman** [comment]
Are you able to provide any information about your System?  What does it contain, what force field does it use, what constraints does it involve?  The best thing would be if you can provide a [serialized System and State](https://openmm.github.io/openmm-cookbook/dev/notebooks/cookbook/Saving%20Systems%20to%20XML%20Files.html) so I can try reproducing it.  Even a list of the forces it contains might provide insight, though.

```python
for f in system.getForces():
  print(f.getName())
```

**Comment by sha256feng** [comment]
It is a membrane protein system with a water box, simulated in CHARMM36m force field, with no constraints. I cannot share that system for some reason, but I am preparing another membrane system to share with you. Thanks.

**Comment by shakespearemorton** [comment]
Hi, I wanted to follow up on this to see if @sha256feng found a reason why there was no improvement with their system? I have recreated the steps here on the HPC system I work with and have had the following results with openmm version 8.0.0.dev-a780005:

1 GPU: 68.03 ns/day 
2 GPU: 87.56 ns/day
gpu: NVIDIA GeForce RTX 3080

The topology between the GPUs is 'SYS' (even though they are on the same node), and I am also trying this on another cluser with PHB connections. You'll have to excuse any ignorance as I am new to working with GPUs, and have only being using OpenMM for a month (LAMMPS leaver). 

However, when I use my custom system there is no increased performance with 2 GPUs when a SYS connection is present, and a performance decrease of about 50% when using PHB. There are quite a few rigid bodies in my system, but few exclusions. I've included the system and a restart file for one of my systems with 142k atoms, but other versions of my system will have upwards of 500-700k. 

[system.zip](https://github.com/openmm/openmm/files/12378904/system.zip)

**Comment by peastman** [comment]
I don't know what SYS and PHB mean?

The two most important factors affecting the speedup with multiple GPUs are the size of the system and the type of connection between the GPUs.  See the cellulose and STMV benchmarks at https://openmm.org/benchmarks.  Cellulose with 408,000 atoms gets a 53% speedup from a second GPU.  STMV with over 1 million atoms gets a 65% speedup.  A system with only 142,000 atoms is probably too small to get much of a speedup.  Your larger systems are better candidates for it.

Those benchmarks were run with A100s connected by NVLink.  That's a very high speed connection.  You're using consumer GPUs that communicate over PCIe, which is much slower.  That will make it difficult to get much speedup with multiple GPUs.
